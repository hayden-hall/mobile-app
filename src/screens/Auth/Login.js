import React, { PureComponent } from 'react';
import {
  View,
  StyleSheet,
  Image,
  } from 'react-native';
import { TextInput, CustomButton, Loader } from '../../components';
import i18n from '../../config/i18n';
import { validateEmail } from '../../utility';
import { login } from '../../services/API/Auth';
import { ASYNC_STORAGE_KEYS } from '../../constants';
import { KeyboardAwareScrollView } from 'react-native-keyboard-aware-scroll-view';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'flex-end',
    marginTop: 20
  },
  logoStyle: { height: 181, width: 181 },
  inputBoxesView: {
    flex: 3,
    width: '90%',
    alignItems: 'center',
    justifyContent: 'center',
    alignSelf: 'center'
  },
  flex1: {
    flex: 1
  },
  flex2: {
    flex: 2
  },
  inputButton: { width: '40%', alignSelf: 'center', paddingTop: 20 }
});

export default class Login extends PureComponent {
  state = {
    emailError: '',
    passwordError: '',
    loginFailError: '',
    email: '',
    password: '',
    isLoading: false
  };

  showsSpinner = (showsSpinner) => {
    this.setState({ isAppLoading: showsSpinner });
  };

  validateInput = () => {
    const { email, password } = this.state;
    if (email.length == 0) {
      this.setState({ emailError: i18n.t('ENTER_EMAIL') });
      return false;
    }
    if (!validateEmail(email)) {
      this.setState({ emailError: i18n.t('ENTER_VALID_EMAIL') });
      return false;
    }
    if (password.length == 0) {
      this.setState({ passwordError: i18n.t('ENTER_PASSWORD') });
      return false;
    }
    this.setState({ emailError: '', passwordError: '' });
    return true;
  };

  makeLogin = async () => {
    if (this.validateInput()) {
      try {
        const { email, password } = this.state;
        this.showsSpinner(true);
        const loginResponse = await login(email, password);
        this.showsSpinner(false);
        if (loginResponse.access_token && loginResponse.instance_url) {
          if (this.props.isLoginModal) {
            this.props.loginSuccessfull();
          } else {
            this.props.navigation.replace('AreaCode', {
              headerTitle: i18n.t('AREA_CODE')
            });
          }
        } else {
          setTimeout(() => {
            alert(`Invalid login`);
          }, 500);
        }
      } catch (error) {
        this.showsSpinner(false);
        setTimeout(() => {
          alert(`${error}`);
        }, 500);
      }
    }
  };

  componentDidMount = () => {
    this.hasLoggedIn();
  };
  
  hasLoggedIn = async () => {
    if (!this.props.isLoginModal) {
      this.showsSpinner(true);
      const CDW_Worked_Id = await storage.load({
        key: ASYNC_STORAGE_KEYS.CDW_WORKED_ID
      }).catch(error => {
        //User not logged in
      });
      this.showsSpinner(false);
      //if user already logged in (= has CDW Id), navigate to survey list
      if (CDW_Worked_Id) {
        this.props.navigation.replace('SurveyList', {
          headerTitle: i18n.t('SURVEYS')
        });
      }
    }
  };

  render() {
    const {
      flex1,
      flex2,
      inputBoxesView,
      container,
      logoStyle,
      inputButton
    } = styles;
    return (
      <KeyboardAwareScrollView style={flex1}>
        <Loader loading={this.state.isLoading} />
        <View style={container}>
          <Image
            source={require('../../../../assets/images/haydenhallicon.png')}
            style={logoStyle}
          />
        </View>
        <View style={inputBoxesView}>
          <TextInput
            onChangeText={email => {
              this.setState({ email });
            }}
            value={this.state.email}
            label={i18n.t('EMAIL')}
            placeholder="john@example.com"
            errorStyle={{ color: 'red' }}
            keyboardType='email-address'
            errorMessage={this.state.emailError}
          />
          <TextInput
            onChangeText={password => {
              this.setState({ password });
            }}
            value={this.state.password}
            label={i18n.t('PASSWORD')}
            placeholder="password"
            errorStyle={{ color: 'red' }}
            secureTextEntry
            errorMessage={this.state.passwordError}
          />
          <View style={inputButton}>
            <CustomButton title={i18n.t('LOGIN')} onPress={this.makeLogin} />
          </View>
        </View>
        <View style={flex2} />
      </KeyboardAwareScrollView>
    );
  }
}
