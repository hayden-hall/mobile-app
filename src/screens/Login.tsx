import React, { useState, useEffect, useContext } from 'react';
import { View, StyleSheet, Image, ImageBackground } from 'react-native';
import { KeyboardAwareScrollView } from 'react-native-keyboard-aware-scroll-view';
import * as SecureStore from 'expo-secure-store';

import { CustomButton, Loader } from '../components';
import { Input } from 'react-native-elements';

import { validateEmail } from '../utility';
import { logger } from '../utility/logger';
import {
  ASYNC_STORAGE_KEYS,
  BACKGROUND_IMAGE_SOURCE,
  BACKGROUND_STYLE,
  BACKGROUND_IMAGE_STYLE,
  APP_FONTS,
  APP_THEME,
} from '../constants';
import LocalizationContext from '../context/localizationContext';

import { authenticate } from '../services/api/auth';
import { notifyError } from '../utility/notification';

export default function Login({ navigation }) {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [emailError, setEmailError] = useState('');
  const [passwordError, setPasswordError] = useState('');
  const [showsSpinner, setShowsSpinner] = useState(false);

  const { t } = useContext(LocalizationContext);

  useEffect(() => {
    hasLoggedIn();
  }, []);

  /**
   * @description Check the user has logged in or not yet.
   */
  const hasLoggedIn = async () => {
    setShowsSpinner(true);

    const cdwWorkedId = await storage
      .load({
        key: ASYNC_STORAGE_KEYS.CDW_WORKED_ID,
      })
      .catch(() => {
        logger('INFO', 'Login', 'User has not logged in yet');
      });
    setShowsSpinner(false);
    //if user already logged in (= has CDW Id), navigate to survey list
    if (cdwWorkedId) {
      logger('INFO', 'Login', `User is already logged in as ${cdwWorkedId}`);
      navigation.navigate('SurveyList');
    }
  };

  const validateInput = () => {
    if (email.length == 0) {
      setEmailError(t('ENTER_EMAIL'));
      return false;
    }
    if (!validateEmail(email)) {
      setEmailError(t('ENTER_VALID_EMAIL'));
      return false;
    }
    if (password.length == 0) {
      setPasswordError(t('ENTER_PASSWORD'));
      return false;
    }
    setEmailError(undefined);
    setPasswordError(undefined);
    return true;
  };

  /**
   * @description Validate form input, retrieve access token, and then navigate to area code screen
   */
  const login = async () => {
    if (!validateInput()) {
      return;
    }
    try {
      setShowsSpinner(true);
      const loginResponse = await authenticate(email, password);
      await SecureStore.setItemAsync('email', email);
      await SecureStore.setItemAsync('password', password);
      logger('DEBUG', 'Login', `${JSON.stringify(loginResponse)}`);
      // TODO: Re-login from survey list screen
      navigation.navigate('AreaCode');
    } catch (error) {
      notifyError(error.error_description);
    } finally {
      setShowsSpinner(false);
    }
  };

  const { flex1, inputBoxesView, container, logoStyle, inputButton, errorStyle, font } = styles;

  return (
    <ImageBackground source={BACKGROUND_IMAGE_SOURCE} style={BACKGROUND_STYLE} imageStyle={BACKGROUND_IMAGE_STYLE}>
      <KeyboardAwareScrollView style={flex1}>
        <Loader loading={showsSpinner} />
        <View style={container}>
          <Image source={require('../../assets/images/haydenhallicon.png')} style={logoStyle} />
        </View>
        <View style={inputBoxesView}>
          <Input
            onChangeText={email => {
              setEmail(email);
            }}
            value={email}
            label={t('EMAIL')}
            placeholder="yourname@example.com"
            leftIcon={{ type: 'material-community', name: 'email-outline', color: APP_THEME.APP_LIGHT_FONT_COLOR }}
            errorStyle={errorStyle}
            labelStyle={font}
            inputStyle={font}
            keyboardType="email-address"
            errorMessage={emailError}
          />
          <Input
            onChangeText={password => {
              setPassword(password);
            }}
            value={password}
            label={t('PASSWORD')}
            placeholder="password"
            leftIcon={{ type: 'material-community', name: 'lock-outline', color: APP_THEME.APP_LIGHT_FONT_COLOR }}
            errorStyle={errorStyle}
            labelStyle={font}
            inputStyle={font}
            secureTextEntry
            errorMessage={passwordError}
          />
          <View style={inputButton}>
            <CustomButton title={t('LOGIN')} onPress={() => login()} />
          </View>
        </View>
      </KeyboardAwareScrollView>
    </ImageBackground>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'flex-end',
    marginTop: 20,
  },
  logoStyle: { height: 181, width: 181 },
  inputBoxesView: {
    flex: 3,
    width: '90%',
    alignItems: 'center',
    justifyContent: 'center',
    alignSelf: 'center',
  },
  flex1: {
    flex: 1,
  },
  flex2: {
    flex: 2,
  },
  font: {
    fontFamily: APP_FONTS.FONT_REGULAR,
  },
  inputButton: { width: '40%', alignSelf: 'center', paddingTop: 20 },
  errorStyle: {
    color: APP_THEME.APP_ERROR_COLOR,
    fontFamily: APP_FONTS.FONT_REGULAR,
  },
});
