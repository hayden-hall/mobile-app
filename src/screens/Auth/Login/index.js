import React, { PureComponent } from 'react';
import {
  View,
  StyleSheet,
  Text,
  Image,
  KeyboardAvoidingView
} from 'react-native';
import { TextInput, CustomButton, Loader } from '../../../components';
import { APP_THEME } from '../../../constants';
import { labels } from '../../../stringConstants';
import { validateEmail } from '../../../utility';
import { login } from '../../../services/API/Auth';
import AsyncStorage from '@react-native-community/async-storage';
import { ASYNC_STORAGE_KEYS } from '../../../constants';
import { KeyboardAwareScrollView } from 'react-native-keyboard-aware-scroll-view';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'flex-end',
    marginTop: 20
  },
  logoStyle: { height: 181, width: 181 },
  inputBoxesView: {
    flex: 3,
    width: '90%',
    alignItems: 'center',
    justifyContent: 'center',
    alignSelf: 'center'
  },
  flex1: {
    flex: 1
  },
  flex2: {
    flex: 2
  },
  inputButton: { width: '40%', alignSelf: 'center', paddingTop: 20 }
});

export default class Login extends PureComponent {
  state = {
    emailError: '',
    passwordError: '',
    loginFailError: '',
    email: '',
    password: '',
    isAppLoading: false
  };

  showHideLoading = show => {
    this.setState({ isAppLoading: show });
  };

  validateInput = () => {
    const { email, password } = this.state;
    if (email.length == 0) {
      this.setState({ emailError: labels.ENTER_EMAIL });
      return false;
    }
    if (!validateEmail(email)) {
      this.setState({ emailError: labels.ENTER_VALID_EMAIL });
      return false;
    }
    if (password.length == 0) {
      this.setState({ passwordError: labels.ENTER_PASSWORD });
      return false;
    }
    this.setState({ emailError: '', passwordError: '' });
    return true;
  };

  makeLogin = async () => {
    if (this.validateInput()) {
      try {
        const { email, password } = this.state;
        this.showHideLoading(true);
        const loginResponse = await login(email, password);
        this.showHideLoading(false);
        if (loginResponse.access_token && loginResponse.instance_url) {
          if (this.props.isLoginModal) {
            this.props.loginSuccessfull();
          } else {
            this.props.navigation.replace('AreaCode', {
              headerTitle: labels.AREA_CODE
            });
          }
        } else {
          setTimeout(() => {
            alert(`Invalid login`);
          }, 500);
        }
      } catch (error) {
        this.showHideLoading(false);
        setTimeout(() => {
          alert(`${error}`);
        }, 500);
      }
    }
  };

  componentDidMount = () => {
    this.checkLogin();
  };

  checkLogin = async () => {
    if (!this.props.isLoginModal) {
      this.showHideLoading(true);
      const CDW_Worked_Id = await AsyncStorage.getItem(
        ASYNC_STORAGE_KEYS.CDW_WORKED_ID
      );
      this.showHideLoading(false);
      if (CDW_Worked_Id) {
        this.props.navigation.replace('SurveyList', {
          headerTitle: labels.SURVEYS
        });
      }
    }
  };

  render() {
    const {
      flex1,
      flex2,
      inputBoxesView,
      container,
      logoStyle,
      inputButton
    } = styles;
    return (
      <KeyboardAwareScrollView style={flex1}>
        <Loader loading={this.state.isAppLoading} />
        <View style={container}>
          <Image
            source={require('../../../../assets/images/haydenhallicon.png')}
            style={logoStyle}
          />
        </View>
        <View style={inputBoxesView}>
          <TextInput
            onChangeText={email => {
              this.setState({ email });
            }}
            value={this.state.email}
            label={labels.EMAIL}
            placeholder="jon@mtxb2b.com"
            errorStyle={{ color: 'red' }}
            keyboardType="email-address"
            errorMessage={this.state.emailError}
          />
          <TextInput
            onChangeText={password => {
              this.setState({ password });
            }}
            value={this.state.password}
            label={labels.PASSWORD}
            placeholder="password"
            errorStyle={{ color: 'red' }}
            secureTextEntry
            errorMessage={this.state.passwordError}
          />
          <View style={inputButton}>
            <CustomButton title={labels.LOGIN} onPress={this.makeLogin} />
          </View>
        </View>
        <View style={flex2} />
      </KeyboardAwareScrollView>
    );
  }
}
