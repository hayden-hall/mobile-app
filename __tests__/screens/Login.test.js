import React from 'react';
import renderer from 'react-test-renderer';

import Login from '../../src/screens/Auth/Login';
import { isExportDeclaration } from 'typescript';

describe('<Login />', () => {
  it('Login', () => {
    const tree = renderer.create(<Login />).toJSON();
    expect(tree.children).toHaveLength(1);
    expect(tree).toMatchSnapshot();
  });
});
