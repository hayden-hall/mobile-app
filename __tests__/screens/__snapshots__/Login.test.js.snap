// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> Login 1`] = `
<RCTScrollView
  automaticallyAdjustContentInsets={false}
  contentInset={
    Object {
      "bottom": 0,
    }
  }
  enableAutomaticScroll={true}
  enableOnAndroid={false}
  enableResetScrollToCoords={true}
  extraHeight={75}
  extraScrollHeight={0}
  getScrollResponder={[Function]}
  handleOnScroll={[Function]}
  keyboardDismissMode="interactive"
  keyboardOpeningTime={250}
  keyboardSpace={0}
  onScroll={[Function]}
  resetKeyboardSpace={[Function]}
  scrollEventThrottle={1}
  scrollForExtraHeightOnAndroid={[Function]}
  scrollIntoView={[Function]}
  scrollToEnd={[Function]}
  scrollToFocusedInput={[Function]}
  scrollToPosition={[Function]}
  showsVerticalScrollIndicator={true}
  style={
    Object {
      "flex": 1,
    }
  }
  update={[Function]}
  viewIsInsideTabBar={false}
>
  <View>
    <Modal
      hardwareAccelerated={false}
      transparent={true}
      visible={true}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#fff",
            "flex": 1,
            "justifyContent": "center",
            "opacity": 0.8,
          }
        }
      >
        <ActivityIndicator
          animating={true}
          color="#01579b"
          hidesWhenStopped={true}
          size="large"
        />
        <Text
          style={
            Object {
              "color": "#01579b",
              "fontFamily": "SalesforceSans-Regular",
              "fontSize": 36,
              "paddingTop": 10,
            }
          }
        >
           
           
        </Text>
      </View>
    </Modal>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "flex-end",
          "marginTop": 20,
        }
      }
    >
      <Image
        source={1}
        style={
          Object {
            "height": 181,
            "width": 181,
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "flex": 3,
          "justifyContent": "center",
          "width": "90%",
        }
      }
    >
      <View
        style={
          Object {
            "padding": 10,
            "paddingHorizontal": 10,
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#706e6b",
              "fontFamily": "SalesforceSans-Regular",
              "fontSize": 14,
              "fontWeight": "bold",
              "marginBottom": 5,
            }
          }
        >
          Email
        </Text>
        <View
          style={
            Object {
              "alignItems": "center",
              "borderBottomWidth": 1,
              "borderColor": "#dedcdb",
              "borderRadius": 4,
              "borderWidth": 1,
              "flexDirection": "row",
              "paddingLeft": 10,
              "transform": Array [
                Object {
                  "translateX": 0,
                },
              ],
            }
          }
        >
          <TextInput
            allowFontScaling={true}
            autoCapitalize="none"
            editable={true}
            keyboardType="email-address"
            multiline={false}
            numberOfLines={1}
            onChangeText={[Function]}
            placeholder="john@example.com"
            rejectResponderTermination={true}
            secureTextEntry={false}
            style={
              Object {
                "alignSelf": "center",
                "color": "black",
                "flex": 1,
                "fontFamily": "SalesforceSans-Regular",
                "fontSize": 16,
                "minHeight": 40,
              }
            }
            testID="RNE__Input__text-input"
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
      </View>
      <View
        style={
          Object {
            "padding": 10,
            "paddingHorizontal": 10,
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#706e6b",
              "fontFamily": "SalesforceSans-Regular",
              "fontSize": 14,
              "fontWeight": "bold",
              "marginBottom": 5,
            }
          }
        >
          Password
        </Text>
        <View
          style={
            Object {
              "alignItems": "center",
              "borderBottomWidth": 1,
              "borderColor": "#dedcdb",
              "borderRadius": 4,
              "borderWidth": 1,
              "flexDirection": "row",
              "paddingLeft": 10,
              "transform": Array [
                Object {
                  "translateX": 0,
                },
              ],
            }
          }
        >
          <TextInput
            allowFontScaling={true}
            autoCapitalize="none"
            editable={true}
            keyboardType="default"
            multiline={false}
            numberOfLines={1}
            onChangeText={[Function]}
            placeholder="password"
            rejectResponderTermination={true}
            secureTextEntry={true}
            style={
              Object {
                "alignSelf": "center",
                "color": "black",
                "flex": 1,
                "fontFamily": "SalesforceSans-Regular",
                "fontSize": 16,
                "minHeight": 40,
              }
            }
            testID="RNE__Input__text-input"
            underlineColorAndroid="transparent"
            value=""
          />
          <View
            style={
              Object {
                "alignItems": "center",
                "height": 40,
                "justifyContent": "center",
                "marginLeft": 15,
              }
            }
          >
            <View>
              <View
                accessible={true}
                focusable={true}
                isTVSelectable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "justifyContent": "center",
                  }
                }
              >
                <Text />
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "alignSelf": "center",
            "paddingTop": 20,
            "width": "40%",
          }
        }
      >
        <View
          style={
            Object {
              "borderRadius": 5,
            }
          }
        >
          <View
            accessible={true}
            focusable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#01579b",
                  "borderColor": "#2089dc",
                  "borderRadius": 5,
                  "borderWidth": 0,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "padding": 8,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "white",
                    "fontFamily": "SalesforceSans-Regular",
                    "fontSize": 16,
                    "paddingBottom": 1,
                    "paddingTop": 2,
                    "textAlign": "center",
                  }
                }
              >
                Log In
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 2,
        }
      }
    />
  </View>
</RCTScrollView>
`;
